File: atividade-24/atividade.go
package main

import "fmt"

type AFD struct {
	state  string
	rotate bool
}

func NewAFD() *AFD {
	return &AFD{
		state:  "q0",
		rotate: false,
	}
}

func (afd *AFD) transition(c rune) {
	switch afd.state {
	case "q0":
		if c == '0' {
			afd.state = "q1"
		}
	case "q1":
		if c == '1' {
			afd.state = "q2"
		} else {
			afd.state = "q0"
		}
	case "q2":
		if c == '0' && !afd.rotate {
			afd.state = "q0"
			afd.rotate = true
		} else if c == '0' && afd.rotate {
			afd.state = "q3" 
		} else if c == '1' {
			afd.state = "q0"
		}
	}
}

func (afd *AFD) isAccepted(input string) bool {
	afd.state = "q0"
	afd.rotate = false
	for _, c := range input {
		afd.transition(c)
	}
	return afd.state == "q3"
}

func main() {
	afd := NewAFD()
	fmt.Println(afd.isAccepted("010010"))  
	fmt.Println(afd.isAccepted("010"))     
	fmt.Println(afd.isAccepted("0101010")) 
	fmt.Println(afd.isAccepted("101010"))  
	fmt.Println(afd.isAccepted("111"))     
	fmt.Println(afd.isAccepted("000"))     
}

File: atividade-24/atividade.py
class AFD:
    def __init__(self):
        self.state = 'q0'
        self.accept_states = {'q3'}
        self.rotate = 0

    def transition(self, char):
        if self.state == 'q0':
            if char == '0':
                self.state = 'q1'
        elif self.state == 'q1':
            if char == '1':
                self.state = 'q2'
            else:
                self.state = 'q0'
        elif self.state == 'q2':
            if char == '0' and self.rotate == 0:
                self.state = 'q0'
                self.rotate = 1
            elif char == '0' and self.rotate == 1:
                self.state = 'q3'
            elif char == '1':
                self.state = 'q0'
        
        

    def is_accepted(self, string):
        self.state = 'q0'
        self.rotate = 0
        for char in string:
            self.transition(char)
        return self.state in self.accept_states

if __name__ == '__main__':
    afd = AFD()
    print(afd.is_accepted("010010"))  
    print(afd.is_accepted("010"))      
    print(afd.is_accepted("0101010"))  
    print(afd.is_accepted("101010"))   
    print(afd.is_accepted("111"))      
    print(afd.is_accepted("000"))      